### PR TITLE
dockerTools: only add "/nix" if it exists

### LINKED ISSUE #14710: dockerTools: failed to remove 'mnt/nix'
I have the following expression:

```nix
with import <nixpkgs> { };

dockerTools.buildImage {
  name = "alpine";
  runAsRoot = ''
    mkdir /lib64 && ln -s /lib/libc.musl-x86_64.so.1 /lib64/ld-linux-x86-64.so.2
    echo "hosts: files dns myhostname mymachines" > /etc/nsswitch.conf
    ln -s /etc/ssl/certs/ca-bundle.crt /etc/ssl/certs/ca-certificates.crt
  '';
  fromImage = dockerTools.buildImage {
    name = "alpine";
    contents = cacert;
    fromImage = dockerTools.pullImage {
      imageName = "alpine";
      imageTag = "3.3";
      imageId = "f58d61a874bedb7cdcb5a409ebb0c53b0656b880695c14e78a69902873358d5f";
      sha256 = "0lvd5zxjgwp3jl5r8qgb2kapmxclpgdv1a7c03yiagxsil5gwb8c";
    };
  };
}
```

When building, I get an error:

```
Formatting '/nix/store/nsvr3yarqmmgx5vjm9vh6ddvbq9hs5xv-docker-layer/disk-image.qcow2', fmt=qcow2 size=1073741824 encryption=off cluster_size=65536 lazy_refcounts=off refcount_bits=16
loading kernel modules...
mounting Nix store...
mounting host's temporary directory...
starting stage 2 (/nix/store/sslrvxb7kwkslbmgwma3wj51qkd1n1yp-vm-run-stage2)
mke2fs 1.42.13 (17-May-2015)
Creating filesystem with 262144 4k blocks and 65536 inodes
Filesystem UUID: ba4455db-f74d-4fed-820d-e5ace1a5eb20
Superblock backups stored on blocks: 
	32768, 98304, 163840, 229376

Allocating group tables: done                            
Writing inode tables: done                            
Writing superblocks and filesystem accounting information: done

Unpacking base image
Unpacking layer 9942f7701da35c4afcf3ce391c045926de5f3547ac9b07341162a439bf135f0d
Unpacking layer f58d61a874bedb7cdcb5a409ebb0c53b0656b880695c14e78a69902873358d5f
rmdir: failed to remove 'mnt/nix': Invalid argument
[    1.524412] reboot: Power down
builder for ‘/nix/store/mygnammwbsjnfbwqmzbhicqwbv2si5zm-docker-layer.drv’ failed with exit code 1
```

I did a bisect and found it's nixpkgs commit 4d200538c22503f900068eb29de6e66a55186260 which broke the expression. @lethalman, could you take a look at this please?

### PR DESCRIPTION
**Pull request by puffnfresh**
The /nix path in 4d200538 of the layer tar didn't exist for some
packages, such as cacert. This is because cacert just creates an /etc
directory and doesn't depend on any other /nix paths. If we tried
putting this directory in the tar and using overlayfs with it, we'd get
"Invalid argument" when trying to remove the directory.

We now check whether the closure is non-empty before telling tar to
store the /nix directory.

Fixes #14710.

### CODE REVIEW

File: pkgs/build-support/docker/default.nix [hunk 1]
```diff
@@ -286,17 +286,20 @@ EOF
...
+          # FIXME: might not be /nix/store
+          echo '/nix' >> layerFiles
+          echo '/nix/store' >> layerFiles
+        fi
+
         echo Adding layer
         tar -tf temp/layer.tar >> baseFiles
         sed 's/^\.//' -i baseFiles
-        comm <(sort -n baseFiles|uniq) <(sort -n layerFiles|uniq|grep -v ${layer}) -1 -3 > newFiles
+        comm <(sort -u baseFiles) <(sort -u layerFiles) -1 -3 > newFiles
```
Comment (lethalman) [inline pkgs/build-support/docker/default.nix]: We must make sure `newFiles` has no duplicates though.

Comment (puffnfresh) [inline pkgs/build-support/docker/default.nix]: I don't think this is any different from before, just a bit more efficient. The `-u` flag will do the `uniq` and I don't think the `-n` flag was doing anything useful. I also swapped the `grep -v` for a prune on the `find`.

But I don't really care about this version, if you prefer the previous version.

Comment (lethalman) [inline pkgs/build-support/docker/default.nix]: Didn't notice the `-u`. But please keep the grep, that prune in the find is unreadable.

**Comment by zimbatm** [comment]
Thanks ! Could be better but it's already an improvement.

**Pull request merged**